site/content/docs/reference/code-walkthrough.md

```diff
@@ -0,0 +1,218 @@
+---
+title: Code Walk-through
+description: A brief overview of the Pinniped source code.
+cascade:
+  layout: docs
+menu:
+  docs:
+    name: Code Walk-through
+    weight: 60
+    parent: reference
+---
+
+## Audience and purpose
+
+The purpose of this document is to provide a high-level, brief introduction to the Pinniped source code for new contributors.
+
+The target audience is someone who wants to read the source code. Users who only want to install and configure Pinniped
+should not need to read this document.
+
+This document aims to help a reader navigate towards the part of the code which they might be interested in exploring in more detail.
+We take an "outside-in" approach, describing how to start finding and understanding code based on several flavors of "outside" entry points.
+
+This document avoids getting too detailed in its description because the details of the code will change over time.
+
+New contributors are also encouraged to read the
+[contributor's guide](https://github.com/vmware-tanzu/pinniped/blob/main/CONTRIBUTING.md)
+and [architecture overview]({{< ref "architecture" >}}).
+
+## Application main functions
+
+There are three binaries in the Pinniped source:
+
+1. The Pinniped CLI
+
+   The `main()` function is in [cmd/pinniped/main.go](https://github.com/vmware-tanzu/pinniped/blob/main/cmd/pinniped/main.go).
+   Each [subcommand]({{< ref "cli" >}}) is in a file:
+   - `pinniped version` in [cmd/pinniped/cmd/version.go](https://github.com/vmware-tanzu/pinniped/blob/main/cmd/pinniped/cmd/version.go)
+   - `pinniped whoami` in [cmd/pinniped/cmd/whoami.go](https://github.com/vmware-tanzu/pinniped/blob/main/cmd/pinniped/cmd/whoami.go)
+   - `pinniped get kubeconfig` in [cmd/pinniped/cmd/kubeconfig.go](https://github.com/vmware-tanzu/pinniped/blob/main/cmd/pinniped/cmd/kubeconfig.go)
+   - The following subcommands are not typically used directly by and end user. Instead, they are usually embedded as
+     a kubectl credential exec plugin in a kubeconfig file:
+     - `pinniped login oidc` in [cmd/pinniped/cmd/login_oidc.go](https://github.com/vmware-tanzu/pinniped/blob/main/cmd/pinniped/cmd/login_oidc.go)
+     - `pinniped login static` in [cmd/pinniped/cmd/login_static.go](https://github.com/vmware-tanzu/pinniped/blob/main/cmd/pinniped/cmd/login_static.go)
+
+2. The Pinniped Kube cert agent component
+
+   The Kube cert agent is a very simple binary that is sometimes deployed by the Pinniped Concierge server component
+   at runtime as a separate Deployment. It exists as a separate binary in the same container image as the other
+   Pinniped server components. When needed, the Concierge will exec into the Deployment's pods to invoke the cert agent
+   binary to query for data. This is to support the Token Credential Request API strategy described in the
+   [Supported Cluster Types document]({{< ref "../reference/supported-clusters" >}}).
```

Comment:
This paragraph is a little vague. Maybe something like, "when needed, the concierge will exec into the kube-cert-agent deployment's pods to query for the keypair used to sign user credentials." rather than just "data" would make it clearer.

Reply:
Updated!